### PR TITLE
Spoiler plugin for text

### PR DESCRIPTION
**Pull request by barboaron**

### CODE REVIEW

File: packages/plugin-spoiler/web/src/spoiler-viewer.jsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    this.state = { styles: mergeStyles({ styles, theme }) };
+  }
+
+  componentWillReceiveProps(nextprops) {
+    const { stateChangeCallBacks } = nextprops;
+    stateChangeCallBacks.push(newState => this.setState(newState));
+  }
+
+  handleClick = event => {
+    !this.state.shouldShowText && event.preventDefault();
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/spoiler-viewer.jsx:31]:
```suggestion
    event.preventDefault();
```

File: packages/plugin-spoiler/web/src/spoiler-viewer.jsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      onClick: this.handleClick,
+      onMouseEnter: this.handleOnMouseEnter,
+      onMouseLeave: this.handleOnMouseLeave,
+      'data-hook': dataHook,
+    };
+    const text = <span {...anchorProps}>{children}</span>;
+    return isMobile || shouldShowText ? (
+      text
+    ) : (
+      <Tooltip content={'Click to reveal'} isTooltipOfSpoiler>
```
Comment (shaulgo) [inline packages/plugin-spoiler/web/src/spoiler-viewer.jsx:59]: translation

File: packages/common/web/src/consts.js [hunk 3]
```diff
@@ -39,6 +39,7 @@ export const IMAGE_TYPE_LEGACY = 'IMAGE';
 export const INDENT_TYPE = 'wix-rich-content-plugin-indent';
 export const LINE_SPACING_TYPE = 'line-spacing';
 export const HEADINGS_DROPDOWN_TYPE = 'HeadingsDropdown';
+export const SPOILER_TYPE = 'SPOILER';
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/common/web/src/consts.js]: done

File: packages/plugin-headings/web/src/toolbar/createToolbar.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ import { decorateComponentWithProps } from 'wix-rich-content-editor-common';
 
 export default settings => ({
   TextButtonMapper: () => ({
-    HeadingsDropdown: {
+    'wix-rich-content-plugin-headings': {
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-headings/web/src/toolbar/createToolbar.js:6]: done

File: packages/plugin-spoiler/web/src/icons/spoilerIcon.svg [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  xmlns="http://www.w3.org/2000/svg"
+>
+  <g stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
+    <g transform="translate(-2144.000000, -748.000000)" fill="currentColor" fillRule="nonzero">
+      <g transform="translate(2144.000000, 748.000000)">
+        <path d="M9.5,2 C10.2187824,2 10.819483,2.50556773 10.9659009,3.18050205 C13.2914419,3.76237692 15,5.69979977 15,8 L15,14 C15,14.5522847 14.5075132,15 13.9,15 L11.9499027,15.0004345 C11.7180997,16.1413382 10.7093254,17 9.5,17 C8.29067455,17 7.28190033,16.1413382 7.05009729,15.0004345 L5.1,15 C4.49248678,15 4,14.5522847 4,14 L4,8 C4,5.69979977 5.7085581,3.76237692 8.03488951,3.17935401 C8.18051695,2.50556773 8.78121764,2 9.5,2 Z M10.9143985,15.000703 L8.08560151,15.000703 C8.29171807,15.5829319 8.8471533,16 9.5,16 C10.1528467,16 10.7082819,15.5829319 10.9143985,15.000703 Z M9.5,4 C7.01471863,4 5,5.78979527 5,7.99761963 L5,14 L14,14 L14,7.99761963 C14,5.78979527 11.9852814,4 9.5,4 Z" />
+      </g>
+    </g>
+  </g>
+</svg>
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/icons/spoilerIcon.svg:15]: done

File: packages/plugin-spoiler/web/src/toolbar/SpoilerButton.jsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  render() {
+    const { theme, isMobile, tabIndex } = this.props;
+    const { isActive } = this.state;
+    return (
+      <InlineToolbarButton
+        onClick={this.handleClick}
+        theme={theme}
+        isMobile={isMobile}
+        tooltipText={'Spoiler'}
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/toolbar/SpoilerButton.jsx:46]: done

File: packages/plugin-spoiler/web/src/spoilerUtilsFn.jsx [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { SPOILER_TYPE } from './types';
+
+export const customStyleFn = styles =>
+  styles.toArray().reduce(cssStyle => {
+    return {
+      ...cssStyle,
+      ...{
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/spoilerUtilsFn.jsx:7]: done

File: packages/editor-common/web/src/Components/Tooltip.jsx [hunk 8]
```diff
@@ -10,6 +13,7 @@ class Tooltip extends React.Component {
     type: PropTypes.oneOf(['success', 'warning', 'error', 'info', 'light', 'dark']),
     shouldRebuildOnUpdate: PropTypes.func,
     place: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
+    isTooltipOfSpoiler: PropTypes.bool,
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/editor-common/web/src/Components/Tooltip.jsx:16]: done

File: packages/plugin-spoiler/web/src/toolbar/createToolbar.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        component: SpoilerButton,
+        isMobile: true,
+        position: { desktop: 0, mobile: 3 },
+        group: {
+          desktop: 3,
+          mobile: 2,
+        },
+      },
+    }),
+    name: 'spoiler',
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/toolbar/createToolbar.js:16]: deleted

File: packages/plugin-spoiler/web/src/spoiler-viewer.jsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  };
+
+  handleOnMouseLeave = () => {
+    this.props.callAllCallbacks({ onHover: false });
+  };
+
+  render() {
+    const { children, isMobile, dataHook } = this.props;
+    const { styles, shouldShowText, onHover } = this.state;
+    const anchorProps = {
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/spoiler-viewer.jsx:46]: done

File: packages/plugin-spoiler/web/src/defaults.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+export const DEFAULTS = {
+  config: {
+    preview: {
+      enable: true,
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/defaults.js:4]: Barak added this for wix/ricos. It's created automatically when creating a new plugin

File: packages/plugin-spoiler/web/src/defaults.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    },
+  },
+};
+export const THEME = colors => ({
+  spoiler: {
+    color: colors.actionColor,
+    textDecoration: 'none',
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/defaults.js:11]: same as above

File: packages/plugin-spoiler/web/src/defaults.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/defaults.js]: I talked with Barak, we don't need this in this plugin - removed

File: packages/plugin-spoiler/web/src/defaults.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/plugin-spoiler/web/src/defaults.js]: removed too

File: packages/viewer/web/src/utils/convertContentState.jsx [hunk 13]
```diff
@@ -157,6 +158,46 @@ const getEntities = (typeMap, pluginProps, styles) => {
   };
 };
 
+const initSpoilers = contentState => {
```
Comment (shaulgo) [inline packages/viewer/web/src/utils/convertContentState.jsx:161]: move to plugin-spoiler

Comment (barboaron) [reply packages/viewer/web/src/utils/convertContentState.jsx]: done

File: CHANGELOG.md [hunk 14]
```diff
@@ -99,6 +99,8 @@ fix bad release in 7.10.1
 
 ## 7.10.1 (Jun 4, 2020)
 ### :rocket: New Feature
+- `spoiler`
```
[the remark this answers is not in the document]
Comment (barboaron) [reply CHANGELOG.md]: done

File: packages/common/web/src/lib/Tooltip.jsx [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    if (!e.target.disabled) {
+      this.timeoutId = setTimeout(() => {
+        this.setState({ tooltipVisible: true });
+      }, 300);
+    }
+  };
+
+  onMouseMove = e => {
+    if (this.props.followMouse) {
+      this.updateTooltipPosition(e.clientX, e.clientY);
```
[the remark this answers is not in the document]
Comment (barboaron) [reply packages/common/web/src/lib/Tooltip.jsx]: fixed

**Pull request merged**